src/shared/uart_framing/uart_framing.hpp

```diff
@@ -0,0 +1,217 @@
+#pragma once
+
+#include <cstdint>
+#include <optional>
+#include <vector>
+
+const uint8_t START_END_FLAG_BYTE = 0x00;
+
+/**
+ * Acts as a frame for uart messages
+ * The type of message should be structs.
+ * These structs should not include pointers or references.
+ *
+ * @tparam T the type of message being framed
+ */
+template <typename T>
+struct UartMessageFrame
+{
+    uint16_t length;
+    uint16_t crc;
+    T message;
+};
+
+namespace
+{
+    /**
+     * Generates a CRC-16 adhering to the CRC-16-CCITT standard
+     *
+     * @param data the data to calculate the crc of
+     * @param length the length of the data
+     * @return CRC-16 for given data and length
+     */
+    template <typename T>
+    uint16_t crc16(const T& data, uint16_t length)
+    {
+        uint8_t x;
+        uint16_t crc = 0xFFFF;
+        auto ptr     = reinterpret_cast<const uint8_t*>(&data);
+        auto bytes   = std::vector<uint8_t>(ptr, ptr + sizeof(data));
+        int i        = 0;
+
+        while (length--)
+        {
+            x = static_cast<uint8_t>(crc >> 8u) ^ bytes[i++];
+            x ^= static_cast<uint8_t>(x >> 4u);
+            crc = static_cast<uint16_t>((crc << 8u) ^ (x << 12u) ^ (x << 5u) ^ x);
+        }
+
+        return crc;
+    }
+    /**
+     * Calculates the length and crc for the given data and returns
+     * a UartMessageFrame with the given data and computed fields
+     *
+     * @tparam T the type of the data param
+     * @param data the data to put in a UartMessageFrame
+     * @return a UartMessageFrame with the given data and computed fields
+     */
+    template <typename T>
+    UartMessageFrame<T> addLengthAndCrc(const T& data)
+    {
+        uint16_t length = sizeof(data);
+        uint16_t crc    = crc16(data, length);
+        UartMessageFrame<T> frame{length, crc, data};
+        return frame;
+    }
```

Comment:
This is creating a new message so maybe just rename this to `createUartMessageFrame`, itll be a UartMessage Frame "factory"